File: packages/core/src/gpu/pipeline.ts
export interface PipelineEntry {
  pipeline: GPUComputePipeline;
  bindGroupLayout: GPUBindGroupLayout;
}

/**
 * Manages creation and caching of WebGPU compute pipelines.
 * Each unique (shader source + entry point) pair produces one cached pipeline.
 */
export class PipelineManager {
  private cache = new Map<string, PipelineEntry>();
  private device: GPUDevice;

  constructor(device: GPUDevice) {
    this.device = device;
  }

  /**
   * Get or create a compute pipeline from WGSL source code.
   * @param key Unique cache key for this pipeline
   * @param wgsl WGSL shader source
   * @param entryPoint Compute entry point name (default: "main")
   * @param constants Pipeline-overridable constants
   */
  getOrCreate(
    key: string,
    wgsl: string,
    entryPoint = "main",
    constants?: Record<string, number>
  ): PipelineEntry {
    const cacheKey = constants
      ? `${key}:${JSON.stringify(constants)}`
      : key;

    const cached = this.cache.get(cacheKey);
    if (cached) return cached;

    const shaderModule = this.device.createShaderModule({ code: wgsl });

    const pipeline = this.device.createComputePipeline({
      layout: "auto",
      compute: {
        module: shaderModule,
        entryPoint,
        constants,
      },
    });

    const bindGroupLayout = pipeline.getBindGroupLayout(0);
    const entry: PipelineEntry = { pipeline, bindGroupLayout };
    this.cache.set(cacheKey, entry);
    return entry;
  }

  /** Clear all cached pipelines */
  clear(): void {
    this.cache.clear();
  }
}

File: packages/rust/crates/oxbitnet/src/gpu/pipeline.rs
use std::collections::HashMap;
use std::sync::Arc;

pub struct PipelineEntry {
    pub pipeline: wgpu::ComputePipeline,
    pub bind_group_layout: wgpu::BindGroupLayout,
}

/// Manages creation and caching of wgpu compute pipelines.
pub struct PipelineManager {
    cache: HashMap<String, Arc<PipelineEntry>>,
    device: Arc<wgpu::Device>,
}

impl PipelineManager {
    pub fn new(device: Arc<wgpu::Device>) -> Self {
        Self {
            cache: HashMap::new(),
            device,
        }
    }

    /// Get or create a compute pipeline from WGSL source code.
    pub fn get_or_create(
        &mut self,
        key: &str,
        wgsl: &str,
        entry_point: &str,
        _constants: Option<&HashMap<String, f64>>,
    ) -> Arc<PipelineEntry> {
        let cache_key = key.to_string();

        if let Some(entry) = self.cache.get(&cache_key) {
            return Arc::clone(entry);
        }

        let shader_module = self.device.create_shader_module(wgpu::ShaderModuleDescriptor {
            label: Some(key),
            source: wgpu::ShaderSource::Wgsl(wgsl.into()),
        });

        let pipeline = self
            .device
            .create_compute_pipeline(&wgpu::ComputePipelineDescriptor {
                label: Some(key),
                layout: None,
                module: &shader_module,
                entry_point: Some(entry_point),
                compilation_options: Default::default(),
                cache: None,
            });

        let bind_group_layout = pipeline.get_bind_group_layout(0);

        let entry = Arc::new(PipelineEntry {
            pipeline,
            bind_group_layout,
        });
        self.cache.insert(cache_key, Arc::clone(&entry));
        entry
    }

    /// Convenience: get_or_create with "main" entry point and no constants.
    pub fn get_or_create_default(&mut self, key: &str, wgsl: &str) -> Arc<PipelineEntry> {
        self.get_or_create(key, wgsl, "main", None)
    }

    pub fn clear(&mut self) {
        self.cache.clear();
    }
}
